File: recipes/subms-health/python/src/subms_health/recipe.py
"""Perf runner emitting the SubMsBenchSummary JSON contract (lang=python).
Stages: register / report / render_json. Run:

    PYTHONPATH="src;../../subms-events/python/src" python -m subms_health.recipe > ../perf/python.json
"""

from __future__ import annotations

import json
import sys
import time

from .component import ComponentHealth
from .env_section import EnvSection, MapEnv
from .registry import HealthConfig, HealthRegistry, RefreshPolicy

INDICATORS = 16
ENV_VARS_PER_SECTION = 8


def _build_registry() -> HealthRegistry:
    reg = HealthRegistry(HealthConfig.sync())
    for i in range(INDICATORS - 2):
        if i % 5 != 0:
            reg.register_fn(
                f"dep-{i}",
                lambda: ComponentHealth.up().with_detail("ping", "ok"),
                RefreshPolicy(interval_ms=1_000),
            )
        else:
            reg.register_fn(
                f"dep-{i}",
                lambda: ComponentHealth.degraded("slow upstream"),
                RefreshPolicy(interval_ms=1_000),
            )
    for s in range(2):
        env = MapEnv()
        for v in range(ENV_VARS_PER_SECTION):
            env.with_(f"KICKSTART_VAR{s}_{v}", f"value-{s}-{v}")
        section = (
            EnvSection(f"deploy-{s}")
            .prefix("KICKSTART_")
            .strip_prefix_in_key(True)
            .lowercase_keys(True)
            .redact_secrets()
        )
        reg.register(section.into_indicator(env), RefreshPolicy(interval_ms=60_000, critical=False))
    return reg


def _summarize(samples):
    s = sorted(samples)
    n = len(s)

    def pct(p):
        return s[min(n - 1, int(p * n))]

    step = max(1, n // 500)
    return {
        "count": n,
        "p50_ns": pct(0.50),
        "p99_ns": pct(0.99),
        "p999_ns": pct(0.999),
        "max_ns": s[-1],
        "mean_ns": int(sum(s) / n),
        "samples_ns": s[::step][:500],
    }


def run(entries: int = 20_000) -> dict:
    register = []
    for _ in range(entries):
        t0 = time.perf_counter_ns()
        reg = _build_registry()
        register.append(time.perf_counter_ns() - t0)

    reg = _build_registry()
    report = []
    for _ in range(entries):
        t0 = time.perf_counter_ns()
        reg.refresh_now()
        report.append(time.perf_counter_ns() - t0)

    reg.refresh_now()
    render = []
    for _ in range(entries):
        t0 = time.perf_counter_ns()
        reg.render()
        render.append(time.perf_counter_ns() - t0)

    return {
        "workload": "subms-health",
        "lang": "python",
        "timestamp": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        "inputs": {"entries": str(entries)},
        "meta": {
            "subms.recipe.slug": "subms-health",
            "subms.recipe.category": "observability",
            "indicators": str(INDICATORS),
            "env_vars_per_section": str(ENV_VARS_PER_SECTION),
            "subms.workload.feature": "cached-snapshot-render",
        },
        "stages": {
            "register": _summarize(register),
            "report": _summarize(report),
            "render_json": _summarize(render),
        },
    }


def main() -> None:
    entries = int(sys.argv[1]) if len(sys.argv) > 1 else 20_000
    json.dump(run(entries), sys.stdout, separators=(",", ":"))


if __name__ == "__main__":
    main()

File: recipes/subms-health/rust/src/recipe.rs
//! `SubMsRecipe` impl. Three stages: `register` (build + register indicators),
//! `report` (force-probe + aggregate + serialise the snapshot), `render_json`
//! (serve the cached snapshot - the per-request hot path).

use std::sync::Arc;

use subms::{SubMsBenchParams, SubMsPerfHarness, SubMsRecipe, SubMsStageKind, SubMsTimer};

use crate::{ComponentHealth, EnvSection, HealthRegistry, MapEnv, RefreshPolicy};

pub struct HealthRecipe;

const INDICATORS: usize = 16;
const ENV_VARS_PER_SECTION: usize = 8;

fn build_registry() -> HealthRegistry {
    let mut reg = HealthRegistry::new();
    for i in 0..(INDICATORS - 2) {
        let name = format!("dep-{i}");
        let healthy = i % 5 != 0;
        reg.register_fn(
            &name,
            RefreshPolicy::new().with_interval_ms(1_000),
            move || {
                if healthy {
                    ComponentHealth::up()
                        .with_detail("ping", "ok")
                        .with_detail("rtt_us", 42i64)
                } else {
                    ComponentHealth::degraded("slow upstream").with_detail("rtt_us", 9_000i64)
                }
            },
        );
    }
    for s in 0..2 {
        let mut env = MapEnv::new();
        for v in 0..ENV_VARS_PER_SECTION {
            env = env.with(&format!("KICKSTART_VAR{s}_{v}"), &format!("value-{s}-{v}"));
        }
        let section = EnvSection::new(&format!("deploy-{s}"))
            .prefix("KICKSTART_")
            .strip_prefix_in_key(true)
            .lowercase_keys(true)
            .redact_secrets();
        reg.register(
            Arc::new(section.into_indicator(Arc::new(env))),
            RefreshPolicy::new()
                .with_interval_ms(60_000)
                .critical(false),
        );
    }
    reg
}

impl SubMsRecipe for HealthRecipe {
    fn name(&self) -> &str {
        "subms-health"
    }

    fn run(&self, h: &mut SubMsPerfHarness, params: &SubMsBenchParams) {
        let entries = params.entries;

        let s_reg = h
            .stage("register", entries)
            .with_kind(SubMsStageKind::BatchOp);
        for _ in 0..entries {
            let t0 = SubMsTimer::tick();
            let reg = build_registry();
            s_reg.record(t0.elapsed_ns());
            std::hint::black_box(&reg);
        }

        let reg = build_registry();
        let s_report = h
            .stage("report", entries)
            .with_kind(SubMsStageKind::HotPath);
        for _ in 0..entries {
            let t0 = SubMsTimer::tick();
            reg.refresh_now();
            s_report.record(t0.elapsed_ns());
        }

        reg.refresh_now();
        let s_render = h
            .stage("render_json", entries)
            .with_kind(SubMsStageKind::HotPath);
        for _ in 0..entries {
            let t0 = SubMsTimer::tick();
            let (code, json) = reg.render();
            s_render.record(t0.elapsed_ns());
            std::hint::black_box((code, json.len()));
        }

        let (_, json) = reg.render();
        h.add_meta("indicators", &INDICATORS.to_string());
        h.add_meta("env_vars_per_section", &ENV_VARS_PER_SECTION.to_string());
        h.add_meta("report_bytes", &json.len().to_string());
        h.add_meta("overall_status", reg.status().as_str());
        h.add_meta("subms.workload.feature", "cached-snapshot-render");
    }
}
